### ISSUE TITLE
Force Redraw after container change

### ISSUE THREAD

**Issue by C00per** [angular-google-chart/angular-google-chart#185]
Hi,
I'm using Bootstrap and perform ng-hide for the first a bootstrap column, so the second column resizes. 
Unfortunately the size of the chart doesn't change. BUT the size is changed if I resize window or change data for the chart.
Is it possible to have a workaround for this? Or at least force chart to redraw with the same data?

**Comment by C00per** [comment]
I've found some kind of workaround by changing binding value
```
       var chart =angular.copy($scope.chart,{});      
       $scope.chart=null;
       $scope.$apply();
        //toggle value
       $scope.openSlider=!$scope.openSlider;        
       $scope.chart=chart;
```
Is there a more elegant solution?

**Comment by Balrog30** [comment]
Just use `ng-if` instead of `ng-hide`.

`ng-hide` just adds a `display: none;` style, which screws up the rendering of the google chart because there is no height to the element at the time that it renders the SVG content.

`ng-if` on the other hand completely removes the directive from the DOM until the expression evaluates to true.

**Comment by C00per** [comment]
Strange, but it is not work. Bootstrap column is removed from DOM but the chart is not changed. Anyway only width has to be changed.

**Comment by Balrog30** [comment]
Can you work with a fixed size?  If you set your chart options for a fixed width and height it should always draw that size.

**Comment by C00per** [comment]
No, I'm want it to be responsible. And it works well except my manipulating with hiding.
It is shame that I cannot force refresh the chart.

**Comment by Balrog30** [comment]
I've done a lot of work on refactoring the main directive for this project into something more extensible.  I'll take a look at this because it has sort of come up before.  As a bit of a hack, you could change one of the chartObject properties (say options, for example), in some inconsequential way, to cause a redraw.  For example:

```JavaScript
$scope.showChart = function(){
  $scope.chartVisible = true;
  $scope.myChart.options.forceRedrawNow = true;
}
```

I'm pretty sure that things in your options object that are not recognized by the Google Charts API will just be ignored, but it would trigger the angular watch on the options object causing angular-google-chart to do a redraw.

**Comment by C00per** [comment]
@Balrog30 thank you for your help.
I've found a workaround. I perfrom `ng-if` on chart after hiding\shwoning another bootstrap column.
```
<div google-chart chart="chart" style="{{chart.cssStyle}}" ng-if="!openSlider"></div>
<div google-chart chart="chart" style="{{chart.cssStyle}}" ng-if="openSlider"></div>
```
so chart always refreshed

**Comment by Balrog30** [comment]
Are you, by chance, using the off-canvas menu, as used in the experimental example [here](http://getbootstrap.com/examples/offcanvas/)? The `openSlider` variable makes me think that you might be.  If you are, then it gives me something to play around with to see why your chart doesn't render properly.  Just having it in an `ng-if` to keep it from trying to render until the container element has dimensions should work... but if it's being animated or something I have absolutely no idea what effect that would have on the chart.

**Comment by C00per** [comment]
@Balrog30 
No, I don't use it. Just simple Bootsrap grid. The left column can be hidden and then the right column changes its size from 11 to 12.

**Comment by Balrog30** [comment]
For the record, I actually had the same sort of thing come up where I needed to force a redraw (in my case I was using the chart's layout interface to overlay something onto the chart in a position relative to the data).  The most elegant solution I found without changing any code in angular-google-chart was something like this:

```JavaScript
function forceChartRedraw(){
    $rootScope.$emit('resizeMsg');
}
```

This basically just tricks angular-google-chart into redrawing as if the window had been resized.  There could be some performance issues if you have a large number of charts with lots of data.  A slightly less elegant solution that forces a redraw of just one chart would be something like this:

```JavaScript
$scope.mychart = {
  // a bunch of other stuff here
  options: {
    forceRedrawCount: 0
  }
};

function forceChartRedraw(){
  $scope.myChart.options.forceRedrawCount++;
}
```

This solution just modifies a watched value on the chart object without actually changing the chart.  This would work well if you have many charts but only want to update one (for performance reasons, for example).

**Comment by C00per** [comment]
I alredy tried the second solution and it doesn't work for me, but the first solution works well. Thank you!
What is the difference between the first solution and my workaround with ng-if? Are there performance problems in using ng-if?

Your solution causes unpleasant blink because chart is re-draws when other elements already changed their size. I think it is possible to play with ng-hide and ng-show to eleminate this blink.

**Comment by Balrog30** [comment]
There's nothing necessarily wrong with your solution. If you do actually have two elements with opposite `ng-if`s, there may be some performance issues from completely re-creating the chartWrappers and charts, if you have a lot of charts or data. Some people were displaying data quantities on the level where performance was an issue.  We used to re-create the chart on every draw, but there were complaints about performance.

The main thing for me was that I needed to redraw the chart based on the arrival of new data, to super-impose a small graphic on the chart.  To get the positions to draw on, I needed the chartWrapper/chart, so I tied the positioning of the graphic to `on-ready`, which only fires on a re-draw.  So, I forced a re-draw on the arrival of new data (no user interaction).  The catch of course being that the new "data", was not in the data table.

It was all very interesting feature to write.  I might write a post about it on my blog when I get a chance.
